broadinstitute/single_cell_portal_core
# Enable exploring custom metrics and one-vs-rest-only DE from authors (SCP-5236)

This lets authors show custom significance metrics and simpler DE analysis outputs in the DE explore UI, .

Previously (#1813), author-uploaded DE had to specify its significance metric as a q-value.  Now, authors can also provide adjusted p-value for significance.  As described in more detail at #1874, adjusted p-values are more common in single-cell differential expression analysis output.

Exploring DE from authors also previously required them to upload files in a more complex format.  That format assumed the author would provide DE that includes "pairwise comparisons" (e.g. A-vs-B).  Work in #1874 enabled simpler DE analysis output for one-vs-rest-_only_ analyses; these simpler analyses are relatively common.  Now, the author DE UI supports exploring such one-vs-rest-only data.

Here's how it looks!

_Explore author DE that has adjusted p-value as its significance metric -- newly possible_

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/6074ad18-9cc0-4982-b1c3-d46607267453

---

_Explore author DE that has q-value as its significance metric -- same as before, no regression_

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/4c80d3f4-77f7-4b16-99be-197709798a9c

---

_Explore author DE that has one-vs-rest-only comparisons -- newly possible_

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/146f67da-3b64-46fb-97e6-51f72031e74b


### Test
1.  If on staging, go to the ["Differential expression" upload tab for SCP358](https://singlecell-staging.broadinstitute.org/single_cell/studies/64ad99c8b0673401ed2e7b1e/upload?tab=differential%20expression).  If local, go to a copy of the human milk DE pilot test
2. Select "Associated clustering: All Cells UMAP" and "Associated annotation: biosample_id"
3. Click "Choose file...", select [seurat_findallmarkers_one-vs-rest_biosample_id.csv](https://github.com/broadinstitute/single_cell_portal_core/files/12768603/seurat_findallmarkers_one-vs-rest_biosample_id.csv)
4.  Click "Upload & Save"
5. Wait 2-3 minutes (after ensuring you're running `rails:jobs work` if local)
6. Go to Explore for the study, select "Clustering: General Celltype" and "Annotation: biosample_id"
7. Click "Differential expression" button, select a group like "B2", confirm you see DE table with slider facet and table column header with "Adj. p-value"
8. If you want to restore SCP-computed DE, go to upload UI, download a cluster file, delete it via UI, then reupload it

If we get [signal users click on "Choose file..." in the "Differential Expression" tab in the upload UI](https://mixpanel.com/s/4bgntR), future work could include tasks like:
* [Enable more custom metric labels in author DE explore UI](https://broadworkbench.atlassian.net/browse/SCP-5352)
* [Surface author DE methods in explore UI](https://broadworkbench.atlassian.net/browse/SCP-5321)

This satisfies SCP-5236.
